src/transport/raw/TCP.cpp

```diff
@@ -290,101 +268,82 @@ CHIP_ERROR TCPBase::SendAfterConnect(const PeerAddress & addr, System::PacketBuf
     return err;
 }
 
-CHIP_ERROR TCPBase::ProcessSingleMessageFromBufferHead(const PeerAddress & peerAddress, const System::PacketBufferHandle & buffer,
-                                                       uint16_t messageSize)
-{
-    CHIP_ERROR err     = CHIP_NO_ERROR;
-    uint8_t * oldStart = buffer->Start();
-    uint16_t oldLength = buffer->DataLength();
-
-    buffer->SetDataLength(messageSize);
-
-    PacketHeader header;
-    err = header.DecodeAndConsume(buffer);
-    SuccessOrExit(err);
-
-    // message receive handler will attempt to free the buffer, however as the buffer may
-    // contain additional data, we retain it to prevent actual free
-    HandleMessageReceived(header, peerAddress, buffer.Retain());
-
-exit:
-    buffer->SetStart(oldStart);
-    buffer->SetDataLength(oldLength);
-
-    return err;
-}
-
 CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const PeerAddress & peerAddress,
                                           System::PacketBufferHandle buffer)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
+    CHIP_ERROR err  = CHIP_NO_ERROR;
+    size_t consumed = 0;
 
     while (!buffer.IsNull())
     {
-        // when a buffer is empty, it can be released back to the app
-        if (buffer->DataLength() == 0)
+        uint8_t messageSizeBuf[kPacketSizeBytes];
+        if (buffer->Read(messageSizeBuf, sizeof(messageSizeBuf)) != CHIP_NO_ERROR)
```

Comment:
It's a little weird to have a mix of the explicit check here but `SuccessOrExit` down below.  I don't really care which one we do that much, in this case, but could we make them consistent?